### PR TITLE
Cypher filtering 1 to 1 relationships

### PR DESCRIPTION
**Pull request by mjfwebb**
# Description

This PR adds the ability to filter on 1 to 1 relationship cypher fields. A contrived example:

Type definitions:
```gql
type Movie
    @node
    @authorization(validate: [{ where: { node: { directed_by: { name: "$jwt.custom_value" } } } }]) {
    title: String
    released: Int
    directed_by: Person!
        @cypher(
            statement: """
            MATCH (this)<-[:DIRECTED]-(director:Person)
            RETURN director
            """
            columnName: "director"
        )
}

type Person @node {
    name: String
    directed: Movie!
        @cypher(
            statement: """
            MATCH (this)-[:DIRECTED]->(movie:Movie)
            RETURN movie
            """
            columnName: "movie"
        )
}
```

You can now query with a filter based on the `directed` field:
```gql
query {
    people(where: { directed: { title: "The Matrix" } }) {
        directed {
            title
            directed_by {
                name
            }
        }
    }
}
```

## Complexity

Medium

# Issue

Relates to #554, but not fully closing as this does not n to n relationship fields.

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [x] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3404 | 3404 | 35 | 54 | 1134 |
|  aggregations.TopLevelAggregateWithMultipleFields | 6802 | 6802 | 42 | 67 | 1134 |
|  aggregations.NestedAggregation | 15407 | 15407 | 53 | 73 | 2174 |
|  aggregations.AggregationWithWhere | 10833 | 10833 | 40 | 64 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20097917 | 20097917 | 2051 | 2053 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20097917 | 20097917 | 1989 | 2028 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 8603 | 8603 | 38 | 38 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8791 | 8791 | 33 | 38 | 2174 |
|  aggregations.DeeplyNestedCount | 10052335 | 10052335 | 2741 | 2816 | 2008534 |
|  aggregations.InterfacesAggregations | 6242 | 6242 | 42 | 52 | 2080 |
|  aggregations.InterfacesAggregationsWithTwoFields | 11444 | 11444 | 72 | 82 | 2080 |
|  batch-create.BatchCreate | 4200 | 4200 | 120 | 164 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 50 | 56 | 11 |
|  connect.createAndConnect | 6433 | 6433 | 156 | 156 | 3003 |
|  connections.Connection | 12951 | 12951 | 59 | 61 | 2174 |
|  connections.NestedConnection | 37705 | 37705 | 137 | 147 | 4516 |
|  create.SimpleMutation | 7 | 7 | 62 | 62 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 29 | 32 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 683 | 762 | 1040 |
|  delete.NestedDeleteInUpdate | 16844 | 16844 | 130 | 147 | 1183 |
|  2871.NestedRelationshipFilter | 19632 | 19632 | 53 | 69 | 4395 |
|  2925.SingleRelationshipFilter | 5245 | 5245 | 38 | 50 | 1040 |
|  2925.NestedSingleRelationshipFilter | 17641 | 17641 | 71 | 98 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 40 | 46 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 56 | 61 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 26 | 29 | 1040 |
|  query.SimpleQueryWithRelationship | 15031 | 15031 | 41 | 41 | 2174 |
|  query.QueryWhere | 8564 | 8564 | 33 | 42 | 2154 |
|  query.SimpleQueryWithNestedWhere | 8713 | 8713 | 46 | 54 | 2154 |
|  query.Nested | 10084891 | 10084891 | 6796 | 6965 | 2008534 |
|  query.NestedWithFilter | 10064992 | 10064992 | 6479 | 6489 | 2004000 |
|  query.OrFilterOnRelationships | 36676 | 36497 | 147 | 234 | 1937 |
|  query.OrFilterOnRelationshipsAndNested | 26749 | 26928 | 183 | 208 | 1937 |
|  query.QueryWithNestedIn | 14132 | 14165 | 102 | 60 | 1844 |
|  query.NestedConnectionWhere | 8703 | 8703 | 48 | 60 | 2174 |
|  query.DeeplyNestedConnectionWhere | 8702 | 8702 | 73 | 90 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 17357 | 17357 | 116 | 151 | 1552 |
|  query.NestedWithRelationshipSingleFilters | 3808 | 3808 | 151 | 172 | 1134 |
|  query.Fulltext | 64 | 64 | 29 | 37 | 16 |
|  query.FulltextWithNestedQuery | 516 | 516 | 41 | 44 | 84 |
|  sorting-and-cypher.TopLevelSortWithCypher | 12961 | 12961 | 38 | 47 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypher | 12961 | 12961 | 57 | 86 | 2174 |
|  sorting-and-cypher.TopLevelSortWithCypherWithNested | 13096 | 13096 | 55 | 67 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithCypherWithNested | 13096 | 13096 | 87 | 100 | 2174 |
|  sorting-and-cypher.TopLevelSortWithExpensiveCypher | 13725 | 13735 | 98 | 134 | 2174 |
|  sorting-and-cypher.TopLevelConnectionSortWithExpensiveCypher | 13266 | 13266 | 97 | 151 | 2174 |
|  sorting.SortMultipleTypes | 3436 | 3436 | 81 | 95 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 13321 | 13321 | 98 | 129 | 2174 |
|  sorting.SortOnNestedFields | 12951 | 12951 | 49 | 59 | 2174 |
|  sorting.SortDeeplyNestedFields | 39785 | 39785 | 80 | 107 | 4516 |
|  sorting.ConnectionWithSort | 3271 | 3271 | 70 | 88 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 64 | 82 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 52 | 76 | 35 |
|  unions.NestedUnion | 309975 | 309975 | 286 | 281 | 33033 |
|  unions.NestedUnionWithMissingFields | 283949 | 283949 | 271 | 266 | 33033 |
|  update.NestedUpdate | 14137 | 14137 | 103 | 111 | 2002 |

</details>
        
Old Schema Generation: 29.453s
Schema Generation: 28.149s
Old Subgraph Schema Generation: 31.038s
Subgraph Schema Generation: 30.835s

File: packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts [hunk 1]
```diff
@@ -133,7 +133,10 @@ export class AttributeAdapter {
 
     isWhereField(): boolean {
         return (
-            (this.typeHelper.isEnum() || this.typeHelper.isSpatial() || this.typeHelper.isScalar()) &&
+            (this.typeHelper.isEnum() ||
+                this.typeHelper.isSpatial() ||
+                this.typeHelper.isScalar() ||
+                (this.isCypher() && Boolean(this.annotations.cypher?.targetEntity))) &&
```
Comment (angrykoala) [inline packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts]: consider moving this new logic to a separate variable, as it is it is quite hard to follow

**Review by angrykoala — APPROVED** [review]

**Comment by angrykoala** [comment]
Maybe worth double check by @MacondoExpress as I've contributed to this PR

File: packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts (hunk 1 above)
Comment (MacondoExpress) [reply packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts]: I agree, it's very hard to understand what is the intention behind each condition.

File: packages/graphql/src/schema-model/attribute/AttributeTypeHelper.ts [hunk 2]
```diff
@@ -177,6 +177,10 @@ export class AttributeTypeHelper {
         return this.type.isRequired;
     }
 
+    public isNullable(): boolean {
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/attribute/AttributeTypeHelper.ts]: I would remove this, don't get me wrong, I have nothing against `isNullable`, but it is just an alias for `isRequired`. I would choose one or the other, with no preference between the two!

File: packages/graphql/src/schema-model/generate-model.ts [hunk 3]
```diff
@@ -127,6 +131,18 @@ function addCompositeEntitiesToConcreteEntity(compositeEntities: CompositeEntity
     });
 }
 
+function hydrateCypherAnnotations(schema: Neo4jGraphQLSchemaModel, concreteEntities: ConcreteEntity[]) {
+    for (const concreteEntity of concreteEntities) {
+        for (const attributeField of concreteEntity.attributes.values()) {
+            if (attributeField.annotations.cypher) {
+                if (attributeField.type instanceof ObjectType) {
+                    attributeField.annotations.cypher.targetEntity = schema.getConcreteEntity(attributeField.type.name);
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/generate-model.ts]: At this moment, if `getConcreteEntity` returns undefined, there is an error. I would catch the error here, rather than propagate the error elsewhere.

File: packages/graphql/src/translate/queryAST/ast/filters/CypherRelationshipFilter.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            return Cypher.head(Cypher.collect(returnVariable));
+        }
+        return returnVariable;
+    }
+
+    private createRelationshipOperation(queryASTContext: QueryASTContext): Cypher.Predicate | undefined {
+        const targetNodePredicates = this.targetNodeFilters.map((c) => c.getPredicate(queryASTContext));
+        const innerPredicate = Cypher.and(...targetNodePredicates);
+
+        switch (this.operator) {
```
Comment (MacondoExpress) [inline packages/graphql/src/translate/queryAST/ast/filters/CypherRelationshipFilter.ts]: I believe here there is no reason to switch depending on the operator.

File: packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts [hunk 5]
```diff
@@ -0,0 +1,1087 @@
...
+                        columnName: "actor"
+                    )
+            }
+
+            type ${Actor} @node {
+                name: String
+                movie: ${Movie}
+                    @cypher(
+                        statement: """
+                        MATCH (this)-[:ACTED_IN]->(movie:${Movie})
```
Comment (MacondoExpress) [inline packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts]: This Cypher could return more than 1 movie, I don't know what we want to do for this case, it could be:
The user needs to verify the correctness of the Cypher statement, in this case, this statement will be wrong.
We handle this case in the library, in this case, we need to test the case when there are more Movies returned or more edges for each movie.

File: packages/graphql/tests/tck/directives/cypher/filtering/cypher-filtering-relationship.test.ts [hunk 6]
```diff
@@ -0,0 +1,1591 @@
...
+            CALL {
+                WITH this
+                CALL {
+                    WITH this
+                    WITH this AS this
+                    MATCH (this)-[:ACTED_IN]->(actor:Actor)
+                    RETURN actor
+                }
+                WITH actor AS this0
+                RETURN this0 AS this1
```
Comment (MacondoExpress) [inline packages/graphql/tests/tck/directives/cypher/filtering/cypher-filtering-relationship.test.ts]: I believe this will change the cardinality when more Actors are returned, worth testing the case I mentioned above!

**Review by MacondoExpress — COMMENTED** [review]
Nice, I left some comments!

File: packages/graphql/src/schema-model/generate-model.ts (hunk 3 above)
Comment (mjfwebb) [reply packages/graphql/src/schema-model/generate-model.ts]: This is a great idea, and I will implement it, but I just want to highlight that `targetEntity` will still remain as an optionally undefined property due to it's limited use case only inside relationship cypher fields.

File: packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts (hunk 5 above)
Comment (mjfwebb) [reply packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts]: Unless I'm misunderstanding, you're pointing out that the test itself is poorly written due to the custom cypher possibly returning more than one movie?

I do think there's something to be said about this particular feature being something more for "power users". We can't be sure that the cypher they enter is correct, which is to say the onus is on the user to provide correct data and correct custom cypher.

Please let me know if I've misunderstood your point.

File: packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts (hunk 1 above)
Comment (mjfwebb) [reply packages/graphql/src/schema-model/attribute/model-adapters/AttributeAdapter.ts]: [I added a new variable](https://github.com/neo4j/graphql/pull/5723/commits/a322af5edaca12c361a8e8d69bd662f65924fac5) named `isCypherRelationshipField`. I'm not entirely sure if this is the best name, but it's at least a bit more readable.

File: packages/graphql/src/schema-model/generate-model.ts (hunk 3 above)
Comment (MacondoExpress) [reply packages/graphql/src/schema-model/generate-model.ts]: Absolutely!

File: packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts (hunk 5 above)
Comment (MacondoExpress) [reply packages/graphql/tests/integration/directives/cypher/filtering/cypher-filtering-relationship.int.test.ts]: Let me rephrase, I believe these tests are wrong or incomplete.
The reason could vary depending on the intention of the feature:
- If we say we cannot handle wrong statements that may return more than a node, then the code should be tested with correct statements (in this case, the statement is wrong as it could return more than a Movie)

- If we say the statement is correct, we should test the behavior when zero or more than a movie is present.

Either way, IMO the cardinality should be handled consistently with the existing behavior, and not depending on whether the target is an Entity or a Scalar or if the field is being used in a filter or a selection

**Review by MacondoExpress — APPROVED** [review]
LGTM!
